File: lib/java/src/main/java/io/github/brenbar/japi/server/Processor.java
package io.github.brenbar.japi.server;

import io.github.brenbar.japi.BinaryEncoder;
import io.github.brenbar.japi.DefaultSerializer;
import io.github.brenbar.japi.Serializer;

import java.util.*;
import java.util.function.Consumer;

public class Processor {

    private Handler handler;
    private Handler internalHandler;
    private Map<String, Object> originalApiDescription;
    private Map<String, Definition> apiDescription;
    private Serializer serializer;
    private Consumer<Throwable> onError;

    private BinaryEncoder binaryEncoder;

    public static class Options {
        private Consumer<Throwable> onError = (e) -> {
        };
        private Serializer serializer = new DefaultSerializer();

        public Options setOnError(Consumer<Throwable> onError) {
            this.onError = onError;
            return this;
        }

        public Options setSerializer(Serializer serializer) {
            this.serializer = serializer;
            return this;
        }
    }

    public Processor(Handler handler, String apiDescriptionJson) {
        this(handler, apiDescriptionJson, new Options());
    }

    public Processor(Handler handler, String apiDescriptionJson, Options options) {
        var description = Parser.newJapi(apiDescriptionJson);
        this.apiDescription = description.parsed();
        this.originalApiDescription = description.original();
        this.serializer = options.serializer;

        var internalDescription = Parser.newJapi(InternalJapi.JSON);

        this.apiDescription.putAll(internalDescription.parsed());
        this.originalApiDescription.putAll(internalDescription.original());

        this.handler = handler;
        this.internalHandler = InternalJapi.build(this.originalApiDescription);
        this.onError = options.onError;

        this.binaryEncoder = BuildBinaryEncoder.build(apiDescription);
    }

    public byte[] process(byte[] inputJapiMessagePayload) {
        return ProcessBytes.process(inputJapiMessagePayload, serializer, onError, binaryEncoder, apiDescription,
                internalHandler, handler);
    }
}

File: lib/py/src/japi/processor.py
from typing import Any, Callable, Dict, List, Optional
from dataclasses import dataclass


class JapiParseError(RuntimeError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


@dataclass
class Options:
    on_error: Callable[[Exception], None] = lambda e: None
    serializer: Serializer = DefaultSerializer()

    def set_on_error(self, on_error: Callable[[Exception], None]) -> 'Options':
        self.on_error = on_error
        return self

    def set_serializer(self, serializer: Serializer) -> 'Options':
        self.serializer = serializer
        return self


class Processor:
    def __init__(self, handler: Handler, api_description_json: str, options: Optional[Options] = None) -> None:
        options = options or Options()
        description = InternalParse.new_japi(api_description_json)
        self.api_description = description.parsed()
        self.original_api_description = description.original()
        self.serializer = options.serializer

        internal_description = InternalParse.new_japi(InternalJapi.JSON)

        self.api_description.update(internal_description.parsed())
        self.original_api_description.update(internal_description.original())

        self.handler = handler
        self.internal_handler = InternalJapi.build(
            self.original_api_description)
        self.on_error = options.on_error

        self.binary_encoder = InternalBinaryEncoderBuilder.build(
            self.api_description)

    def process(self, input_japi_message_payload: bytes) -> bytes:
        return InternalProcess.process(
            input_japi_message_payload,
            self.serializer,
            self.on_error,
            self.binary_encoder,
            self.api_description,
            self.internal_handler,
            self.handler
        )
